source/utils/exceptions.cpp

```diff
@@ -81,8 +81,8 @@ invalid_data_type::~invalid_data_type()
 {
 }
 
-invalid_file::invalid_file(const std::string &filename)
-    : exception(std::string("couldn't open file: (") + filename + ")")
+invalid_file::invalid_file(const std::string &reason)
+    : exception(std::string("couldn't open file, reason given: (") + reason + ")")
```

Comment:
Using both a colon and parentheses to separate the reason seems superfluous to me.